front/lib/auth.ts

```diff
@@ -429,77 +437,75 @@ export async function getAPIKey(
   return new Ok(key);
 }
 
-const DEFAULT_DATASOURCES_COUNT_LIMIT = 1;
-const DEFAULT_DATASOURCES_DOCUMENTS_COUNT_LIMIT = 32;
-const DEFAULT_DATASOURCES_DOCUMENTS_SIZE_MB_LIMIT = 1;
-
 /**
  * Construct the PlanType for the provided workspace.
  * @param w WorkspaceType the workspace to get the plan for
  * @returns PlanType
  */
-export function planForWorkspace(w: Workspace): PlanType {
-  const limits = {
-    dataSources: {
-      count: DEFAULT_DATASOURCES_COUNT_LIMIT,
-      documents: {
-        count: DEFAULT_DATASOURCES_DOCUMENTS_COUNT_LIMIT,
-        sizeMb: DEFAULT_DATASOURCES_DOCUMENTS_SIZE_MB_LIMIT,
-      },
-      managed: false,
-    },
-    largeModels: false,
-  };
+export async function planForWorkspace(
+  w: Workspace
+): Promise<Promise<PlanType>> {
+  let activeSubscription: Subscription | null = null;
+  const today = new Date();
+  today.setHours(0, 0, 0, 0);
+
+  activeSubscription = await Subscription.findOne({
+    where: { workspaceId: w.id, status: "active" },
+  });
 
-  if (w.plan) {
-    let plan = {} as any;
-    try {
-      plan = JSON.parse(w.plan) as any;
-    } catch (e) {
-      logger.error({ planJSON: w.plan, error: e }, "Error parsing plan JSON");
-    }
+  // Default values when no subscription
+  let plan: PlanAttributes = FREE_TEST_PLAN_DATA;
+  let startDate = today;
+  let endDate = null;
 
-    if (plan.limits) {
-      if (plan.limits.largeModels) {
-        if (typeof plan.limits.largeModels === "boolean") {
-          limits.largeModels = plan.limits.largeModels;
-        }
-      }
-      if (plan.limits.dataSources) {
-        if (
-          plan.limits.dataSources.count &&
-          typeof plan.limits.dataSources.count === "number"
-        ) {
-          limits.dataSources.count = plan.limits.dataSources.count;
-        }
-        if (plan.limits.dataSources.documents) {
-          if (
-            plan.limits.dataSources.documents.count &&
-            typeof plan.limits.dataSources.documents.count === "number"
-          ) {
-            limits.dataSources.documents.count =
-              plan.limits.dataSources.documents.count;
-          }
-          if (
-            plan.limits.dataSources.documents.sizeMb &&
-            typeof plan.limits.dataSources.documents.sizeMb === "number"
-          ) {
-            limits.dataSources.documents.sizeMb =
-              plan.limits.dataSources.documents.sizeMb;
-          }
-        }
-        if (
-          plan.limits.dataSources.managed &&
-          typeof plan.limits.dataSources.managed === "boolean"
-        ) {
-          limits.dataSources.managed = plan.limits.dataSources.managed;
-        }
-      }
+  if (activeSubscription) {
+    const subscribedPlan = await Plan.findOne({
```

Comment:
With a join I have this: 
```
Query: Executed (default): 
  SELECT 
    "subscription"."id",
    "subscription"."createdAt",
    "subscription"."updatedAt",
    "subscription"."sId",
    "subscription"."status",
    "subscription"."startDate",
    "subscription"."endDate",
    "subscription"."planId",
    "subscription"."workspaceId",
    "plan"."id" AS "plan.id",
    "plan"."createdAt" AS "plan.createdAt",
    "plan"."updatedAt" AS "plan.updatedAt",
    "plan"."code" AS "plan.code",
    "plan"."name" AS "plan.name",
    "plan"."maxMessages" AS "plan.maxMessages",
    "plan"."maxUsersInWorkspace" AS "plan.maxUsersInWorkspace",
    "plan"."isSlackbotAllowed" AS "plan.isSlackbotAllowed",
    "plan"."isManagedSlackAllowed" AS "plan.isManagedSlackAllowed",
    "plan"."isManagedNotionAllowed" AS "plan.isManagedNotionAllowed",
    "plan"."isManagedGoogleDriveAllowed" AS "plan.isManagedGoogleDriveAllowed",
    "plan"."isManagedGithubAllowed" AS "plan.isManagedGithubAllowed",
    "plan"."maxNbStaticDataSources" AS "plan.maxNbStaticDataSources",
    "plan"."maxNbStaticDocuments" AS "plan.maxNbStaticDocuments",
    "plan"."maxSizeStaticDataSources" AS "plan.maxSizeStaticDataSources"
  FROM 
    "subscriptions" AS "subscription" 
  LEFT OUTER JOIN 
    "plans" AS "plan" 
  ON 
    "subscription"."planId" = "plan"."id" 
  WHERE 
    "subscription"."workspaceId" = 1 
  AND 
    "subscription"."status" = 'active' 
  LIMIT 1;

Execution time: 8 ms
```

I'll check if I can do better.

Reply:
ok I added limitation on attributes and indexes.
```
Query: Executed (default): 
  SELECT 
    "subscription"."id",
    "subscription"."startDate",
    "subscription"."endDate",
    "plan"."id" AS "plan.id",
    "plan"."code" AS "plan.code",
    "plan"."name" AS "plan.name",
    "plan"."isSlackbotAllowed" AS "plan.isSlackbotAllowed",
    "plan"."maxMessages" AS "plan.maxMessages",
    "plan"."isManagedSlackAllowed" AS "plan.isManagedSlackAllowed",
    "plan"."isManagedNotionAllowed" AS "plan.isManagedNotionAllowed",
    "plan"."isManagedGoogleDriveAllowed" AS "plan.isManagedGoogleDriveAllowed",
    "plan"."isManagedGithubAllowed" AS "plan.isManagedGithubAllowed",
    "plan"."maxNbStaticDataSources" AS "plan.maxNbStaticDataSources",
    "plan"."maxNbStaticDocuments" AS "plan.maxNbStaticDocuments",
    "plan"."maxSizeStaticDataSources" AS "plan.maxSizeStaticDataSources",
    "plan"."maxUsersInWorkspace" AS "plan.maxUsersInWorkspace"
  FROM 
    "subscriptions" AS "subscription"
  LEFT OUTER JOIN 
    "plans" AS "plan"
  ON 
    "subscription"."planId" = "plan"."id"
  WHERE 
    "subscription"."workspaceId" = 1
  AND 
    "subscription"."status" = 'active'
  LIMIT 1;

Execution time: 2 ms
```
It's not always 2ms but it improves the initial version.

Reply:
Added a "required true" to have an INNER JOIN instead of the LEFT OUTER JOIN

```
Query: Executed (default): 
  SELECT 
    "subscription"."id",
    "subscription"."startDate",
    "subscription"."endDate",
    "plan"."id" AS "plan.id",
    "plan"."code" AS "plan.code",
    "plan"."name" AS "plan.name",
    "plan"."isSlackbotAllowed" AS "plan.isSlackbotAllowed",
    "plan"."maxMessages" AS "plan.maxMessages",
    "plan"."isManagedSlackAllowed" AS "plan.isManagedSlackAllowed",
    "plan"."isManagedNotionAllowed" AS "plan.isManagedNotionAllowed",
    "plan"."isManagedGoogleDriveAllowed" AS "plan.isManagedGoogleDriveAllowed",
    "plan"."isManagedGithubAllowed" AS "plan.isManagedGithubAllowed",
    "plan"."maxNbStaticDataSources" AS "plan.maxNbStaticDataSources",
    "plan"."maxNbStaticDocuments" AS "plan.maxNbStaticDocuments",
    "plan"."maxSizeStaticDataSources" AS "plan.maxSizeStaticDataSources",
    "plan"."maxUsersInWorkspace" AS "plan.maxUsersInWorkspace"
  FROM 
    "subscriptions" AS "subscription"
  INNER JOIN 
    "plans" AS "plan"
  ON 
    "subscription"."planId" = "plan"."id"
  WHERE 
    "subscription"."workspaceId" = 1
  AND 
    "subscription"."status" = 'active'
  LIMIT 1;
```

Reply:
This looks like a reasonable query. Can we explain it and explain the two separate query as per IRL? :pray:

Reply:
### Joined query: 
```sql
  EXPLAIN SELECT 
    "subscription"."id",
    "subscription"."startDate",
    "subscription"."endDate",
    "plan"."id" AS "plan.id",
    "plan"."code" AS "plan.code",
    "plan"."name" AS "plan.name",
    "plan"."isSlackbotAllowed" AS "plan.isSlackbotAllowed",
    "plan"."maxMessages" AS "plan.maxMessages",
    "plan"."isManagedSlackAllowed" AS "plan.isManagedSlackAllowed",
    "plan"."isManagedNotionAllowed" AS "plan.isManagedNotionAllowed",
    "plan"."isManagedGoogleDriveAllowed" AS "plan.isManagedGoogleDriveAllowed",
    "plan"."isManagedGithubAllowed" AS "plan.isManagedGithubAllowed",
    "plan"."maxNbStaticDataSources" AS "plan.maxNbStaticDataSources",
    "plan"."maxNbStaticDocuments" AS "plan.maxNbStaticDocuments",
    "plan"."maxSizeStaticDataSources" AS "plan.maxSizeStaticDataSources",
    "plan"."maxUsersInWorkspace" AS "plan.maxUsersInWorkspace"
  FROM 
    "subscriptions" AS "subscription"
  INNER JOIN 
    "plans" AS "plan"
  ON 
    "subscription"."planId" = "plan"."id"
  WHERE 
    "subscription"."workspaceId" = 1
  AND 
    "subscription"."status" = 'active'
  LIMIT 1;
```

```
Limit  (cost=0.27..16.53 rows=1 width=1081)
  ->  Nested Loop  (cost=0.27..16.53 rows=1 width=1081)
        ->  Index Scan using subscriptions_status on subscriptions subscription  (cost=0.13..8.15 rows=1 width=24)
"              Index Cond: ((status)::text = 'active'::text)"
"              Filter: (""workspaceId"" = 1)"
        ->  Index Scan using plans_pkey on plans plan  (cost=0.14..8.16 rows=1 width=1061)
"              Index Cond: (id = subscription.""planId"")"
```

### Sequential queries: 

```sql
EXPLAIN
SELECT
	"id",
	"startDate",
	"endDate",
	"planId"
FROM
	"subscriptions" AS "subscription"
WHERE
	"subscription"."workspaceId" = 1
	AND "subscription"."status" = 'active'
LIMIT 1;
```
```
Limit  (cost=0.13..8.15 rows=1 width=24)
  ->  Index Scan using subscriptions_status on subscriptions subscription  (cost=0.13..8.15 rows=1 width=24)
"        Index Cond: ((status)::text = 'active'::text)"
"        Filter: (""workspaceId"" = 1)"
```


```sql
SELECT
	"code",
	"name",
	"isSlackbotAllowed",
	"maxMessages",
	"isManagedSlackAllowed",
	"isManagedNotionAllowed",
	"isManagedGoogleDriveAllowed",
	"isManagedGithubAllowed",
	"maxNbStaticDataSources",
	"maxNbStaticDocuments",
	"maxSizeStaticDataSources",
	"maxUsersInWorkspace"
FROM
	"plans" AS "plan"
WHERE
	"plan"."id" = 1;
```

```
Index Scan using plans_pkey on plans plan  (cost=0.14..8.16 rows=1 width=1057)
  Index Cond: (id = 1)
```